src/execution/execute.ts

```diff
@@ -795,63 +797,63 @@ function completeLeafValue(
 function completeAbstractValue(
   exeContext: ExecutionContext,
   returnType: GraphQLAbstractType,
+  abstractType: GraphQLAbstractType,
   fieldNodes: ReadonlyArray<FieldNode>,
   info: GraphQLResolveInfo,
   path: Path,
   result: unknown,
 ): PromiseOrValue<ObjMap<unknown>> {
-  const resolveTypeFn = returnType.resolveType ?? exeContext.typeResolver;
+  const resolveTypeFn = abstractType.resolveType ?? exeContext.typeResolver;
   const contextValue = exeContext.contextValue;
-  const runtimeType = resolveTypeFn(result, contextValue, info, returnType);
-
-  if (isPromise(runtimeType)) {
-    return runtimeType.then((resolvedRuntimeType) =>
-      completeObjectValue(
-        exeContext,
-        ensureValidRuntimeType(
-          resolvedRuntimeType,
-          exeContext,
-          returnType,
-          fieldNodes,
-          info,
-          result,
-        ),
-        fieldNodes,
-        info,
-        path,
-        result,
-      ),
-    );
-  }
+  const runtimeTypeName = resolveTypeFn(
+    result,
+    contextValue,
+    info,
+    abstractType,
+  );
 
-  return completeObjectValue(
+  return resolveRuntimeTypeName(
     exeContext,
-    ensureValidRuntimeType(
-      runtimeType,
-      exeContext,
-      returnType,
-      fieldNodes,
-      info,
-      result,
-    ),
+    returnType,
+    abstractType,
+    runtimeTypeName,
     fieldNodes,
     info,
     path,
     result,
   );
 }
 
-function ensureValidRuntimeType(
-  runtimeTypeName: unknown,
+function resolveRuntimeTypeName(
```

Comment:
This function now essentially completes, suggest renaming to completeAbstractValueImpl or something else.

Reply:
i pushed a change renaming it